lib/iris/coords.py

```diff
@@ -2921,6 +2921,24 @@ def xml_element(self, doc):
         return element
 
 
+_dim_coord_cache = {}
+
+
+def dim_coord_from_regular(*args, **kwargs):
+    coord_system = kwargs.pop("coord_system", None)
```

Comment:
`coord_system` is a bit more complicated  because it can't be hashed, hence not just using `lru_cache` here

Reply:
I think this probably needs a bit more documentation in the code. a comment or docstring about why you're doing it.

Reply:
Alternatively, could you implement `__hash__` on coord_system objects?

Reply:
Given that a `coord_system` is mutable, I don't think I should implement `__hash__`? Then again, I shouldn't rely on the coord system that I stored not being mutated. I'll see what seems best for resolving that

Reply:
Now using the repr, which given what I know of coord_systems seems trustworthy

Reply:
Might need to put an `@abstractmethod` __repr__ on the CoordSystem base class (is that possible on a dunder method?) to ensure that all custom CoordSystems conform to this

Reply:
Done